modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/logical/IgniteMultiJoinOptimizeBushyRule.java

```diff
@@ -0,0 +1,453 @@
+ * from best plans, until all relations are joined.
+ *
+ * <p>Current limitations are as follow:<ol>
+ *     <li>Only INNER joins are supported</li>
+ *     <li>Number of relations to optimize is limited to 20. This is due to time and memory complexity of algorithm chosen.</li>
+ *     <li>Disjunctive predicate is not considered as connections.</li>
+ * </ol>
+ */
+@Value.Enclosing
+public class IgniteMultiJoinOptimizeBushyRule
+        extends RelRule<IgniteMultiJoinOptimizeBushyRule.Config>
+        implements TransformationRule {
+
+    private static final int MAX_JOIN_SIZE = 20;
+
+    /**
+     * Comparator that puts better vertexes first.
+     *
+     * <p>Better vertex is the one that incorporate more relations, or costs less.
+     */
+    private static final Comparator<Vertex> VERTEX_COMPARATOR = 
+            Comparator.<Vertex>comparingInt(v -> bitCount(v.id))
+                    .reversed()
+                    .thenComparingDouble(v -> v.cost);
+
+    /** Creates a MultiJoinOptimizeBushyRule. */
+    private IgniteMultiJoinOptimizeBushyRule(Config config) {
+        super(config);
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        MultiJoin multiJoinRel = call.rel(0);
+
+        int numberOfRelations = multiJoinRel.getInputs().size();
+        if (numberOfRelations > MAX_JOIN_SIZE) {
+            return;
+        }
+
+        // Currently, algorithm below can handle only INNER JOINs
+        if (multiJoinRel.isFullOuterJoin()) {
+            return;
+        }
+
+        for (JoinRelType joinType : multiJoinRel.getJoinTypes()) {
+            if (joinType != JoinRelType.INNER) {
+                return;
+            }
+        }
+
+        LoptMultiJoin multiJoin = new LoptMultiJoin(multiJoinRel);
+
+        RexBuilder rexBuilder = multiJoinRel.getCluster().getRexBuilder();
+        RelBuilder relBuilder = call.builder();
+        RelMetadataQuery mq = call.getMetadataQuery();
+
+        List<RexNode> unusedConditions = new ArrayList<>();
+
+        Int2ObjectMap<List<Edge>> edges = collectEdges(multiJoin, unusedConditions);
+        Int2ObjectMap<Vertex> bestPlan = new Int2ObjectOpenHashMap<>();
+        BitSet connections = new BitSet(1 << numberOfRelations);
+
+        int id = 0b1;
+        int fieldOffset = 0;
+        for (RelNode input : multiJoinRel.getInputs()) {
+            TargetMapping mapping = Mappings.offsetSource(
+                    Mappings.createIdentity(input.getRowType().getFieldCount()),
+                    fieldOffset,
+                    multiJoin.getNumTotalFields()
+            );
+
+            bestPlan.put(id, new Vertex(id, mq.getRowCount(input), input, mapping));
+            connections.set(id);
+
+            id <<= 1;
+            fieldOffset += input.getRowType().getFieldCount();
+        }
+
+        Vertex bestSoFar = null;
+        for (int s = 0b11; s < 1 << numberOfRelations; s++) {
```

Comment:
let's introduce constant instead of  magic number

Reply:
I believe this "magic number" is self-explanatory: it represents a bit mask with two lowermost bit set to 1

Reply:
Magic numbers are often self-explanatory, however, best practices are introducing constants for them

Reply:
could you suggest the name for the new constant then? Also, should we introduce constant for `1` and `null`?

Reply:
MIN_SUBSET_RELATIONS ?

Reply:
`MIN_SUBSET` what does it mean? If you refer to the size of subset, then it's not quite correct, because there are several options for subset of size 2: 0b11, 0b101, 0b110, , 0b1010, etc. The crucial part is that we use two least significant bits here